Add examples to API that support a query. (#1544)
Hyperfoil/Horreum: docs/site/content/en/docs/Reference/APIFilterQueryReference.md

---

> title: Horreum API Filter Query Reference

Reviewer: ```suggestion
title: API Filter Queries
```

Reviewer: The Title is used in the menu, "Horreum" and "Reference" are duplicated information in that context

description: Reference guide for API Filter Query
date: 2024-03-27
weight: 3
---

> In Horreum the API provides the ability to filter results. Used when HTTP API response content type is `application-json`. The `API Filter Query` operates on the server side to reduce payload size.

Reviewer: ```suggestion
Some APIs provide the ability to filter results when HTTP API response content type is `application-json`. The `API Filter Query` operates on the server side to reduce payload size.
```

API operations that support Filter Query provide an HTTP request parameter named *query*. The value of the *query* parameter is a `JSONPath` expression. Any syntax errors in the `JSONPath` will cause the operation to fail with details of the  syntax issue.

To use the Filter Query with an operation that returns a [Run](/docs/concepts/core-concepts/#run) data can be done as follows.

```bash
curl -s  'http://localhost:8080/api/sql/1/queryrun?query=$.*&array=true'  -H 'content-type: application/json' -H 'Authorization: Bearer '$TOKEN

> {"valid":true,"jsonpath":"$.*","errorCode":0,"sqlState":null,"reason":null,"sql":null,"value":"[\"urn:acme:benchmark:0.1\", [{\"test\": \"Foo\", \"duration\": 10, \"requests\": 123}, {\"test\": \"Bar\", \"duration\": 20, \"requests\": 456}], \"defec8eddeadbeafcafebabeb16b00b5\", \"This gets lost by the transformer\"]"}

Reviewer: Can the output JSON be formatted so it is split over multiple lines? atm it is a long single line in the rendered doc

Author: sure

Author: I expanded the outer block of JSON. Do you want the Inner block also expanded ?

Reviewer: I think it is ok to leave it as a string, as the output is not a valid JSON node, but a string represenation

```

The response contains an enclosing JSON object that is common for Filter Query operations. The *value* property contains an array. The elements of the array contain the values of properties in the uploaded [Run](/docs/concepts/core-concepts/#run) JSON. Note the additional HTTP request query parameter named *array*. Setting *array* is necessary to return dynamically sized results.

To filter the above API operation to retrieve only the *results* property of the uploaded [Run](/docs/concepts/core-concepts/#run) object the *query* parameter is defined as `$.results`

```bash
curl -s  'http://localhost:8080/api/sql/1/queryrun?query=$.results&array=true'  -H 'content-type: application/json' -H 'Authorization: Bearer '$TOKEN

> {"valid":true,"jsonpath":"$.results","errorCode":0,"sqlState":null,"reason":null,"sql":null,"value":"[[{\"test\": \"Foo\", \"duration\": 10, \"requests\": 123}, {\"test\": \"Bar\", \"duration\": 20, \"requests\": 456}]]"}

Reviewer: same as above re multi-line json

Author: sure
